virttest/qemu_vm.py

```diff
@@ -1038,9 +1038,36 @@ def add_memorys(devices, params):
                 options.append("maxmem=%s" % mem_params["maxmem"])
                 if mem_params.get("slots"):
                     options.append("slots=%s" % mem_params["slots"])
-            for name in params.objects("mem_devs"):
-                dev = devices.memory_define_by_params(params, name)
-                devs.extend(dev)
+            if params.get("vm_sgx_epc_devs"):
+                sgx_epc_memdev_id = dict()
+                for name in params.objects("mem_devs"):
+                    dev = devices.memory_define_by_params(params, name)
+                    if dev[0].params["backend"] == "memory-backend-epc":
+                        sgx_epc_memdev_id[name] = dev[0].params["id"]
+                    devs.extend(dev)
+                opt = "sgx-epc"
+                machine_type = params.get("machine_type")
+                machine_help = devices.execute_qemu("-M %s,?" % machine_type)
+                if re.search(r"%s=" % opt, machine_help, re.MULTILINE):
+                    sgx_epc_cmd = ""
+                    for idx, ele in enumerate(params.objects(
+                            "vm_sgx_epc_devs")):
+                        memdev_name = params.get("vm_sgx_epc_memdev_%s" % ele)
+                        numa_id = params.get("vm_sgx_epc_node_%s" % ele)
+                        sgx_epc_cmd += ",{opt}.{id}.memdev={epc_id}," \
+                                       "{opt}.{id}.node={numa_id}".format(
+                                        opt=opt, id=idx,
+                                        epc_id=sgx_epc_memdev_id[memdev_name],
+                                        numa_id=numa_id)
```

Comment:
```suggestion
                        opt_mapping = {
                                'memdev': params.get("vm_sgx_epc_memdev_%s" % ele),
                                'node': params.get("vm_sgx_epc_node_%s" % ele)
                                }
                        for opt, val in opt_mapping.items():
                            if val:
                                sgx_epc_cmd += ",{cap_type}.{idx}.{opt}={val}".format(
                                        cap_type=cap_type, idx=idx, opt=opt, val=val)

```

Reply:
Thanks Yongxue, I can change the command line assignment here to opt_mapping, but in opt_mapping memdev value is not params.get("vm_sgx_epc_memdev_%s" % ele), this value gets from the device id, and should be sgx_epc_memdev_id[params.get("vm_sgx_epc_memdev_%s" % ele)].